content/blog/2023-12-18-torcx-deprecation-for-systemd-sysext.md

```diff
@@ -0,0 +1,82 @@
+
+Also, the Flatcar docs have been recommending [downloading plain docker / containerd binaries](https://www.flatcar.org/docs/latest/container-runtimes/use-a-custom-docker-or-containerd-version/) instead of using torcx for customising docker for quite a while.
+
+For almost two years, torcx was only used in Flatcar to ship a single docker version - with all of the drawbacks listed above.
+With the advent of systemd-sysext, we will deprecate torcx and remove it from Flatcar in early 2024.
+
+
+**Sysext to the rescue**
+
+If you want to try our new sysext-based composed images yourself, have a look at the latest Alpha releases - these don't ship torcx anymore, and include docker and containerd as sysexts instead.
+
+Sysexts have the following advantages:
+- Do not require changes in the base OS image.
+  Sysexts can ship arbitrary applications and libraries, and will integrate seamlessly into the base OS via overlay mounts.
+- Are easy to integrate in the build system as binaries do not need special handling.
+- Apply / merge faster into the OS via overlayfs, and do not consume additional memory for storage.
+- Can be updated independently of the base OS by use of systemd-sysupdate.
+  This way, custom docker installations can be kept up to date by users without the need to re-provision nodes.
+  (Note that docker and containerd shipped with the base OS will continue to be updated as part of the base OS.)
+- Both containerd and docker can be disabled via ignition configs simply by overriding respective symlinks from `/etc/extensions/...`.
+  This can be useful when shipping a custom container runtime (or in cases where docker is simply not used).
+  When disabled, the respective binaries and libraries are not merged, and are missing from the OS filesystem.
+
+Starting with [Alpha-3794.0.0](https://www.flatcar.org/releases#release-3794.0.0), containerd and docker are included in the base OS image as sysexts.
+This means that in the regular Flatcar base OS image, there are no `containerd` or `docker` binaries in `/usr/bin` (where formerly the torcx wrapper scripts resided) - check out the [OS image contents](https://alpha.release.flatcar-linux.net/amd64-usr/current/flatcar_production_image_contents.txt) yourself!
+Instead, you'll find `containerd-flatcar.raw` and `docker-flatcar.raw` in `/usr/share/flatcar/sysext/`, and respective symlinks in `/usr/share/flatcar/etc/extensions/` (which is merged into `/etc/` at boot).
+
+
+**That's all well and good, but what does that mean for Flatcar users?**
+
+We aim for a seamless transition and have invested into significant test coverage to ensure user workloads are not disrupted by the change.
+Users will continue to enjoy a robust container OS experience, and do not need to do anything to adopt to the new system.
```

Comment:
unless they're currently using torcx (which we believe very few, if any users, are doing today).

Reply:
We have a script to convert torcx images: https://www.flatcar.org/docs/latest/provisioning/sysext/#converting-a-torcx-image

We also have docs for passing a sysext image in: https://www.flatcar.org/docs/latest/provisioning/sysext/#supplying-your-sysext-image-from-ignition